power_grid_model_c/power_grid_model/include/power_grid_model/main_core/output.hpp

```diff
@@ -98,351 +98,335 @@ constexpr ResIt produce_output(MainModelState<ComponentContainer> const& state,
-          std::forward_iterator ResIt>
+template <std::derived_from<Fault> Component, class ComponentContainer, short_circuit_math_output_type MathOutputType>
     requires model_component_state_c<MainModelState, ComponentContainer, Component> &&
              model_component_state_c<MainModelState, ComponentContainer, Node>
-constexpr ResIt output_result(MainModelState<ComponentContainer> const& state,
-                              std::vector<MathOutputType> const& math_output, ResIt res_it) {
-    return detail::produce_output<Component, Idx2D>(
-        state, res_it, [&state, &math_output](Fault const& fault, Idx2D math_id) {
-            if (math_id.group == -1) {
-                return fault.get_null_sc_output();
-            }
+inline auto output_result(Component const& fault, MainModelState<ComponentContainer> const& state,
+                          std::vector<MathOutputType> const& math_output, Idx2D math_id) {
+    if (math_id.group == -1) {
+        return fault.get_null_sc_output();
+    }
 
-            auto const u_rated = state.components.template get_item<Node>(fault.get_fault_object()).u_rated();
-            return fault.get_sc_output(math_output[math_id.group].fault[math_id.pos], u_rated);
-        });
+    auto const u_rated = state.components.template get_item<Node>(fault.get_fault_object()).u_rated();
+    return fault.get_sc_output(math_output[math_id.group].fault[math_id.pos], u_rated);
 }
 
 // output transformer tap regulator
 template <std::derived_from<TransformerTapRegulator> Component, class ComponentContainer,
-          steady_state_math_output_type MathOutputType, std::forward_iterator ResIt>
+          steady_state_math_output_type MathOutputType>
+    requires model_component_state_c<MainModelState, ComponentContainer, Component>
+constexpr auto output_result(Component const& /* transformer_tap_regulator */,
+                             MainModelState<ComponentContainer> const& /* state */,
+                             std::vector<MathOutputType> const& /* math_output */, Idx const /* obj_seq */) {
+    // TODO: this function is not implemented
+    using sym = typename MathOutputType::sym;
+    return typename TransformerTapRegulator::OutputType<sym>{};
+}
+template <std::derived_from<TransformerTapRegulator> Component, class ComponentContainer,
+          short_circuit_math_output_type MathOutputType>
     requires model_component_state_c<MainModelState, ComponentContainer, Component>
+inline auto output_result(Component const& transformer_tap_regulator,
+                          MainModelState<ComponentContainer> const& /* state */,
+                          std::vector<MathOutputType> const& /* math_output */, Idx const /* obj_seq */) {
+    return transformer_tap_regulator.get_null_sc_output();
+}
+
+// output base component
+template <std::derived_from<Base> Component, class ComponentContainer, math_output_type MathOutputType,
+          std::forward_iterator ResIt>
```

Comment:
This might not be the first occurance, but maybe:
```suggestion
          std::forward_iterator ResIter>
```
?

Reply:
this is a convention also used fairly often in the STL. I prefer not to change that in this PR if you're OK with that